### ISSUE TITLE
Consider hand-written JIT helpers for more platforms

### ISSUE THREAD

**Issue by AaronRobinsonMSFT** [dotnet/runtime#67776]
There are some JIT helpers that are hand-written assembly but are only available on Windows-x64. The impact of these helpers can be profound, which means we might be missing benefits on non-Windows-x64 platforms.

JIT helpers as assembly with fallback to VM (C++), Windows-x64 only.
|                    Method |      Mean |     Error |    StdDev |
|-------------------------- |----------:|----------:|----------:|
|     CORINFO_HELP_NEWSFAST |  1.865 ns | 0.0356 ns | 0.0333 ns |
|          CORINFO_HELP_BOX |  4.234 ns | 0.0396 ns | 0.0351 ns |
|  CORINFO_HELP_NEWARR_1_VC | 17.932 ns | 0.0935 ns | 0.0781 ns |
| CORINFO_HELP_NEWARR_1_OBJ |  9.989 ns | 0.2062 ns | 0.1828 ns |

JIT helpers from the VM (C++), All other platforms. Collected on Windows-x64 for relative performance impact.
|                    Method |     Mean |    Error |   StdDev |
|-------------------------- |---------:|---------:|---------:|
|     CORINFO_HELP_NEWSFAST | 12.84 ns | 0.309 ns | 0.402 ns |
|          CORINFO_HELP_BOX | 15.50 ns | 0.316 ns | 0.363 ns |
|  CORINFO_HELP_NEWARR_1_VC | 30.37 ns | 0.243 ns | 0.190 ns |
| CORINFO_HELP_NEWARR_1_OBJ | 23.83 ns | 0.262 ns | 0.219 ns |

It could be possible to write the JIT helpers with new intrinsics as IL instead of assembly on all the platforms.

See the dynamic update for JIT helper:

https://github.com/dotnet/runtime/blob/ff6abfea0bb9b71eb3215157749047dd120a4d9f/src/coreclr/vm/jitinterfacegen.cpp#L86-L90

/cc @davidwrighton

**Comment by davidwrighton** [comment]
I'd really like to avoid hand-coded assembly for these things. I don't think the performance of these scenarios is quite important enough to justify hand coded assembly on all of our architectures, but if we could write C# (even C# which required some magic intrinsics to work) it would probably be worth it.

**Comment by davidwrighton** [comment]
Could you attach a copy of the benchmark you're using to get these numbers?

**Comment by AaronRobinsonMSFT** [comment]
Oops. Looks like I missed the dynamic options used on non-Windows. See

https://github.com/dotnet/runtime/blob/ff6abfea0bb9b71eb3215157749047dd120a4d9f/src/coreclr/vm/jitinterfacegen.cpp#L76-L79

Let me try them on Windows and compare that as well.

**Comment by EgorBo** [comment]
> CORINFO_HELP_BOX

From my understanding this helper is only used for unoptimzed code and JIT inserts a `CORINFO_HELP_NEWSFAST` instead (in contrast with `CORINFO_HELP_BOX_NULLABLE`)

**Comment by EgorBo** [comment]
So I am looking at `JIT_NewS_MP_FastPortable` and it seems it can be easily converted to C# but might require an intrinsic to get current `allocContext` (double indirect load basically to a global location) but I am not sure what exactly it solves?
I assume in theory we can make it inlineable for hot paths? (and e.g. emit object size as a constant)

Is the problem that C++ version of it emits worse code that the assembly version?

**Comment by EgorBo** [comment]
Something like this I assume:
![image](https://user-images.githubusercontent.com/523221/162577674-a1356f3c-586a-422e-9070-98b4440406eb.png)

**Comment by jkotas** [comment]
You would also need to make sure that the whole method is non-interruptible. We cannot take a GC in the middle of the allocation helper.

**Comment by AaronRobinsonMSFT** [comment]
> You would also need to make sure that the whole method is non-interruptible. 

Right. This was why I alluded to "new intrinsics" for a helper in IL. Do we have a mechanism already to indicate a managed method is non-interruptible?

**Comment by jkotas** [comment]
> Do we have a mechanism already to indicate a managed method is non-interruptible?

We do not have a way to indicate in IL that the method has to be non-interruptible.

**Comment by mangod9** [comment]
Recently we have noticed that `JIT_NewS_MP_FastPortable` helper using more cycles on arm64 when compared to x64. Is there any optimization which might be missing on arm64 here?

**Comment by jkotas** [comment]
@mangod9 What is the OS that you are see it on?

The `JIT_New` helpers are often the first piece of code that touches new memory and tend to incur a lot of cache misses. If you are seeing more cycles spent in these helpers on Arm64, it may be due to higher cost of cache misses. Are you able to tell how much cycles are spent due to cache misses in these helpers?

**Comment by mangod9** [comment]
this is on linux, will try to determine if we can get any perf counters related to cache misses.

**Comment by kunalspathak** [comment]
related: https://github.com/dotnet/runtime/issues/99552
